From 9b9cabdc6e49a7e0879c3bae3d529f108a3dc6ab Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Fri, 28 Dec 2018 22:01:04 +0100
Subject: [PATCH 1/1] efi_loader: provide test files for bootmgr

For testing the 'bootefi bootmgr' sub-command we need multiple UEFI
applications.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 cmd/Kconfig               | 13 ++++++-------
 lib/efi_loader/Makefile   |  9 ++++++++-
 lib/efi_loader/boot0001.c | 28 ++++++++++++++++++++++++++++
 lib/efi_loader/boot0002.c | 28 ++++++++++++++++++++++++++++
 4 files changed, 70 insertions(+), 8 deletions(-)
 create mode 100644 lib/efi_loader/boot0001.c
 create mode 100644 lib/efi_loader/boot0002.c

diff --git a/cmd/Kconfig b/cmd/Kconfig
index 308e5458b9a..db05b60b7b9 100644
--- a/cmd/Kconfig
+++ b/cmd/Kconfig
@@ -224,22 +224,21 @@ config CMD_BOOTEFI
 	help
 	  Boot an EFI image from memory.
 
-config CMD_BOOTEFI_HELLO_COMPILE
-	bool "Compile a standard EFI hello world binary for testing"
+config CMD_BOOTEFI_TEST_FILES
+	bool "Build EFI binaries for testing"
 	depends on CMD_BOOTEFI && (ARM || X86 || RISCV)
 	default y
 	help
-	  This compiles a standard EFI hello world application with U-Boot so
-	  that it can be used with the test/py testing framework. This is useful
-	  for testing that EFI is working at a basic level, and for bringing
-	  up EFI support on a new architecture.
+	  This compiles multiple EFI binaries that can be used with the test/py
+	  testing framework. This is useful for testing that EFI is working at
+	  a basic level, and for bringing up EFI support on a new architecture.
 
 	  No additional space will be required in the resulting U-Boot binary
 	  when this option is enabled.
 
 config CMD_BOOTEFI_HELLO
 	bool "Allow booting a standard EFI hello world for testing"
-	depends on CMD_BOOTEFI_HELLO_COMPILE
+	depends on CMD_BOOTEFI_TEST_FILES
 	help
 	  This adds a standard EFI hello world application to U-Boot so that
 	  it can be used with the 'bootefi hello' command. This is useful
diff --git a/lib/efi_loader/Makefile b/lib/efi_loader/Makefile
index ac37cf6ece3..5d14385f264 100644
--- a/lib/efi_loader/Makefile
+++ b/lib/efi_loader/Makefile
@@ -9,11 +9,18 @@
 CFLAGS_efi_boottime.o += \
   -DFW_VERSION="0x$(VERSION)" \
   -DFW_PATCHLEVEL="0x$(PATCHLEVEL)"
+
+CFLAGS_boot0001.o := $(CFLAGS_EFI) -ffreestanding
+CFLAGS_boot0002.o := $(CFLAGS_EFI) -ffreestanding
 CFLAGS_helloworld.o := $(CFLAGS_EFI) -Os -ffreestanding
+CFLAGS_REMOVE_boot0001.o := $(CFLAGS_NON_EFI)
+CFLAGS_REMOVE_boot0002.o := $(CFLAGS_NON_EFI)
 CFLAGS_REMOVE_helloworld.o := $(CFLAGS_NON_EFI)
 
-ifneq ($(CONFIG_CMD_BOOTEFI_HELLO_COMPILE),)
+ifneq ($(CONFIG_CMD_BOOTEFI_TEST_FILES),)
 always += helloworld.efi
+always += boot0001.efi
+always += boot0002.efi
 endif
 
 obj-$(CONFIG_CMD_BOOTEFI_HELLO) += helloworld_efi.o
diff --git a/lib/efi_loader/boot0001.c b/lib/efi_loader/boot0001.c
new file mode 100644
index 00000000000..75f5973a188
--- /dev/null
+++ b/lib/efi_loader/boot0001.c
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Test file for boot manager.
+ *
+ * Copyright (c) 2018 Heinrich Schuchardt <xypron.glpk@gmx.de>
+ *
+ * This UEFI binary just outputs its name.
+ */
+
+#include <efi_api.h>
+
+/**
+ * efi_main() - entry point of the EFI application.
+ *
+ * @handle:	handle of the loaded image
+ * @systable:	system table
+ * @return:	status code
+ */
+efi_status_t EFIAPI efi_main(efi_handle_t handle,
+			     struct efi_system_table *systable)
+{
+	/* UEFI requires CR LF */
+	systable->con_out->output_string(con_out, L"Executing BOOT0001\r\n");
+	systable->boottime->exit(handle, ret, 0, NULL);
+
+	/* We should never arrive here */
+	return ret;
+}
diff --git a/lib/efi_loader/boot0002.c b/lib/efi_loader/boot0002.c
new file mode 100644
index 00000000000..245144bc993
--- /dev/null
+++ b/lib/efi_loader/boot0002.c
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Test file for boot manager.
+ *
+ * Copyright (c) 2018 Heinrich Schuchardt <xypron.glpk@gmx.de>
+ *
+ * This UEFI binary just outputs its name.
+ */
+
+#include <efi_api.h>
+
+/**
+ * efi_main() - entry point of the EFI application.
+ *
+ * @handle:	handle of the loaded image
+ * @systable:	system table
+ * @return:	status code
+ */
+efi_status_t EFIAPI efi_main(efi_handle_t handle,
+			     struct efi_system_table *systable)
+{
+	/* UEFI requires CR LF */
+	systable->con_out->output_string(con_out, L"Executing BOOT0002\r\n");
+	systable->boottime->exit(handle, ret, 0, NULL);
+
+	/* We should never arrive here */
+	return ret;
+}
-- 
2.19.2

